src/OrbitGl/ThreadTrack.h

```diff
@@ -56,19 +57,25 @@ class ThreadTrack final : public TimerTrack {
 
   [[nodiscard]] Color GetTimerColor(const orbit_client_protos::TimerInfo& timer, bool is_selected,
                                     bool is_highlighted) const override;
-  void SetTimesliceText(const orbit_client_protos::TimerInfo& timer, double elapsed_us, float min_x,
-                        float z_offset, TextBox* text_box) override;
+  [[nodiscard]] Color GetTimerColor(const TextBox& text_box, const internal::DrawData& draw_data);
+  void SetTimesliceText(const orbit_client_protos::TimerInfo& timer, float min_x, float z_offset,
+                        TextBox* text_box) override;
   [[nodiscard]] std::string GetBoxTooltip(const Batcher& batcher, PickingId id) const override;
 
   [[nodiscard]] float GetHeight() const override;
   [[nodiscard]] float GetHeaderHeight() const override;
 
   void UpdatePositionOfSubtracks();
+  void UpdatePrimitivesOfSubtracks(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                   PickingMode picking_mode, float z_offset);
   void UpdateMinMaxTimestamps();
 
   std::shared_ptr<orbit_gl::ThreadStateBar> thread_state_bar_;
   std::shared_ptr<orbit_gl::CallstackThreadBar> event_bar_;
   std::shared_ptr<orbit_gl::TracepointThreadBar> tracepoint_bar_;
+
+  absl::Mutex scope_tree_mutex_;
+  ScopeTree<TextBox> scope_tree_;
```

Comment:
We should start using `GUARDED_BY` annotations for new mutex guarded elements: https://abseil.io/docs/cpp/guides/synchronization#annotation-guidelines

Reply:
Interesting, yeah I think we should look into it. I feel I need to understand it a bit deeper before starting applying it however.